packages/iframe-execution-environment-service/src/IframeExecutionEnvironmentService.ts

```diff
@@ -0,0 +1,337 @@
+import { Duplex } from 'stream';
+import { nanoid } from 'nanoid';
+import pump from 'pump';
+import { ObservableStore } from '@metamask/obs-store';
+import ObjectMultiplex from '@metamask/object-multiplex';
+import { WindowPostMessageStream } from '@metamask/post-message-stream';
+import { PLUGIN_STREAM_NAMES } from '@mm-snap/workers';
+import { createStreamMiddleware } from 'json-rpc-middleware-stream';
+import { PluginData } from '@mm-snap/types';
+import {
+  JsonRpcEngine,
+  JsonRpcRequest,
+  PendingJsonRpcResponse,
+} from 'json-rpc-engine';
+import { ExecutionEnvironmentService } from '@mm-snap/controllers';
+
+export type SetupPluginProvider = (pluginName: string, stream: Duplex) => void;
+
+interface JobControllerArgs {
+  setupPluginProvider: SetupPluginProvider;
+  iframeUrl: URL;
+}
+
+interface JobStreams {
+  command: Duplex;
+  rpc: Duplex | null;
+  _connection: WindowPostMessageStream;
+}
+
+// The plugin is the callee
+export type PluginRpcHook = (
+  origin: string,
+  request: Record<string, unknown>,
+) => Promise<unknown>;
+
+interface EnvMetadata {
+  id: string;
+  streams: JobStreams;
+  rpcEngine: JsonRpcEngine;
+}
+
+export class IframeExecutionEnvironmentService
+  implements ExecutionEnvironmentService
+{
+  public store: ObservableStore<{ jobs: Record<string, EnvMetadata> }>;
+
+  private _pluginRpcHooks: Map<string, PluginRpcHook>;
+
+  public _iframeWindow?: Window;
+
+  public iframeUrl: URL;
+
+  private jobs: Map<string, EnvMetadata>;
+
+  private setupPluginProvider: SetupPluginProvider;
+
+  private pluginToJobMap: Map<string, string>;
+
+  private jobToPluginMap: Map<string, string>;
+
+  constructor({ setupPluginProvider, iframeUrl }: JobControllerArgs) {
+    this.iframeUrl = iframeUrl;
+    this.setupPluginProvider = setupPluginProvider;
+    this.store = new ObservableStore({ jobs: {} });
+    this.jobs = new Map();
+    this.pluginToJobMap = new Map();
+    this.jobToPluginMap = new Map();
+    this._pluginRpcHooks = new Map();
+  }
+
+  private _setJob(jobId: string, jobWrapper: EnvMetadata): void {
+    this.jobs.set(jobId, jobWrapper);
+
+    const newJobState = {
+      ...(this.store.getState().jobs as Record<string, EnvMetadata>),
+      [jobId]: jobWrapper,
+    };
+    this.store.updateState({ jobs: newJobState });
+  }
+
+  private _deleteJob(jobId: string): void {
+    this.jobs.delete(jobId);
+
+    const newJobState = {
+      ...(this.store.getState().jobs as Record<string, EnvMetadata>),
+    };
+    delete newJobState[jobId];
+    this.store.updateState({ jobs: newJobState });
+  }
+
+  private async _command(
+    jobId: string,
+    message: JsonRpcRequest<unknown>,
+  ): Promise<unknown> {
+    if (typeof message !== 'object') {
+      throw new Error('Must send object.');
+    }
+
+    const jobWrapper = this.jobs.get(jobId);
+    if (!jobWrapper) {
+      throw new Error(`Job with id ${jobId} not found.`);
```

Comment:
Mark and I usually do this for template string values, in case they are missing/empty, and also to highlight that they are variables:
```suggestion
      throw new Error(`Job with id "${jobId}" not found.`);
```